File: tests/oracle/typefactory_codeflags_decode_1204.cc
/*
 * TYPEFACTORY-CODEFLAGS-DECODE-0001: locked Ghidra 12.0.4 oracle for
 * TypeFactory::decodeTypeWithCodeFlags (type.cc:4193-4212), the
 * decodeCode stub/compare/define chain it calls (type.cc:4401-4429,
 * 2903-2931), and the XmlDecode cursor partial state every error path
 * leaves behind.
 *
 * The locked oracle's decodeTypeWithCodeFlags never reaches decodeCode's
 * prototype handling on real streams: decodeBasic runs the outer element's
 * attribute enumeration to exhaustion, the WORDSIZE loop (type.cc:4201-4207,
 * no rewindAttributes) reads nothing, and decodeCode's decodeStub then
 * re-reads the SAME still-open element and raises "Bad size for type ".
 * The fixture records that behaviour verbatim across code-pointer flag
 * combinations (varargs/model/ctor/dtor/thiscall models), the two earlier
 * error paths, the decodeCode paths reachable through decodeType
 * (prototype-less stub creation, completion via the setPrototype wrapper
 * with a null prototype, dedup identity, redefinition and name-clash
 * errors with their surviving factory state), and the pointer->code chain.
 */
#include "bfd_arch.hh"
#include "libdecomp.hh"
#include "type.hh"

#include <iostream>
#include <sstream>
#include <stdexcept>
#include <string>
#include <vector>

namespace {

using namespace ghidra;

/// Run one decodeTypeWithCodeFlags probe over the wrapped inner XML and
/// emit the error record plus the cursor partial state: the still-open
/// element's unread children (peek), the successful close of that element,
/// and the readable <void/> sibling behind it.
static void runCodeFlagsCase(TypeFactory *factory,Architecture *glb,
                              const string &key,const string &innerXml,
                              bool isConstructor,bool isDestructor)
{
  const string xml = "<root>" + innerXml + "<void/></root>";
  std::istringstream stream(xml);
  DocumentStorage store;
  Document *doc = store.parseDocument(stream);
  XmlDecode decoder(glb,doc->getRoot());
  uint4 rootId = decoder.openElement();
  bool thrown = false;
  string text;
  try {
    factory->decodeTypeWithCodeFlags(decoder,isConstructor,isDestructor);
  }
  catch(const LowlevelError &err) {
    thrown = true;
    text = err.explain;
  }
  std::cout << key << ".thrown=" << (thrown ? 1 : 0) << '\n';
  std::cout << key << ".text=" << text << '\n';
  std::cout << key << ".peek=" << (decoder.peekElement() != 0 ? 1 : 0) << '\n';
  decoder.closeElementSkipping(0);
  int4 resume = 0;
  try {
    uint4 vId = decoder.openElement(ELEM_VOID);
    decoder.closeElement(vId);
    resume = 1;
  }
  catch(DecoderError &err) { }
  std::cout << key << ".resume_void=" << resume << '\n';
  decoder.closeElement(rootId);
}

/// Decode one inner <type> element through TypeFactory::decodeType and
/// return the resulting Datatype, or null with errText set.
static Datatype *decodeTypeStr(TypeFactory *factory,Architecture *glb,
                               const string &innerXml,string &errText)
{
  const string xml = "<root>" + innerXml + "</root>";
  std::istringstream stream(xml);
  DocumentStorage store;
  Document *doc = store.parseDocument(stream);
  XmlDecode decoder(glb,doc->getRoot());
  uint4 rootId = decoder.openElement();
  Datatype *res = (Datatype *)0;
  errText.clear();
  try {
    res = factory->decodeType(decoder);
  }
  catch(const LowlevelError &err) {
    errText = err.explain;
  }
  decoder.closeElement(rootId);
  return res;
}

static void emitThrown(const string &key,bool thrown,const string &text)

{
  std::cout << key << ".thrown=" << (thrown ? 1 : 0) << '\n';
  std::cout << key << ".text=" << text << '\n';
}

static void emitCodeType(const string &key,const Datatype *ct)

{
  std::cout << key << ".name=" << ct->getName() << '\n';
  std::cout << key << ".id=" << ct->getId() << '\n';
  std::cout << key << ".size=" << ct->getSize() << '\n';
  std::cout << key << ".meta=" << static_cast<int4>(ct->getMetatype()) << '\n';
  std::cout << key << ".incomplete=" << (ct->isIncomplete() ? 1 : 0) << '\n';
  std::cout << key << ".varlength=" << (ct->isVariableLength() ? 1 : 0) << '\n';
  const TypeCode *tc = dynamic_cast<const TypeCode *>(ct);
  std::cout << key << ".proto=" << ((tc != (const TypeCode *)0 && tc->getPrototype() != (const FuncProto *)0) ? 1 : 0) << '\n';
}

void runFixture(const string &specDirectory,const string &binary)

{
  vector<string> specPaths;
  specPaths.push_back(specDirectory);
  startDecompilerLibrary(specPaths);
  BfdArchitecture architecture(binary,"default",&std::cerr);
  DocumentStorage store;
  architecture.init(store);
  if (architecture.archid != "x86:LE:64:default:gcc")
    throw std::runtime_error("runtime architecture/compiler drifted: " + architecture.archid);

  TypeFactory *factory = architecture.types;
  if (factory == (TypeFactory *)0)
    throw std::runtime_error("architecture has no TypeFactory");

  // ---------------- Wave 1: decodeTypeWithCodeFlags ----------------
  // a_*: the nested pointer->code form under every code-pointer flag
  // combination. The locked oracle throws "Bad size for type " (empty
  // name) from decodeCode's decodeStub re-read of the exhausted outer
  // element before ANY flag is consulted; the records prove the flag
  // indifference byte-for-byte.
  runCodeFlagsCase(factory,&architecture,"w1.a_plain",
    "<type metatype=\"ptr\" size=\"8\">"
    "<type metatype=\"code\" size=\"1\"/>"
    "</type>", true, false);
  runCodeFlagsCase(factory,&architecture,"w1.a_varargs",
    "<type metatype=\"ptr\" size=\"8\">"
    "<type metatype=\"code\" size=\"1\">"
    "<prototype model=\"__stdcall\" dotdotdot=\"true\"><returnsym/></prototype>"
    "</type>"
    "</type>", false, true);
  runCodeFlagsCase(factory,&architecture,"w1.a_model",
    "<type metatype=\"ptr\" size=\"8\">"
    "<type metatype=\"code\" size=\"1\">"
    "<prototype model=\"unknown_cc\" constructor=\"true\" destructor=\"true\"/>"
    "</type>"
    "</type>", true, true);
  runCodeFlagsCase(factory,&architecture,"w1.a_thiscall",
    "<type metatype=\"ptr\" size=\"8\" wordsize=\"1\">"
    "<type metatype=\"code\" size=\"1\">"
    "<prototype model=\"__thiscall\"><returnsym/></prototype>"
    "</type>"
    "</type>", false, false);
  // b_*: metatype check failures before any child is read.
  runCodeFlagsCase(factory,&architecture,"w1.b_code",
    "<type metatype=\"code\" size=\"1\"/>", true, false);
  runCodeFlagsCase(factory,&architecture,"w1.b_unspec",
    "<type size=\"4\"/>", true, true);
  // c_*: the first decodeBasic's own Bad size throw (named and anonymous).
  runCodeFlagsCase(factory,&architecture,"w1.c_nosize_named",
    "<type metatype=\"ptr\" name=\"vp\"/>", true, false);
  runCodeFlagsCase(factory,&architecture,"w1.c_nosize_anon",
    "<type metatype=\"ptr\"/>", false, true);

  // ---------------- Wave 2: decodeCode via decodeType ----------------
  // d_named: prototype-less code stub decoded, completed in place by the
  // setPrototype wrapper (Ghidra clears type_incomplete even for a null
  // prototype), observable through the returned handle.
  string err;
  Datatype *cfOne = decodeTypeStr(factory,&architecture,
    "<type metatype=\"code\" name=\"cf_one\" size=\"1\"/>", err);
  std::cout << "w2.d_named.thrown=" << (cfOne == (Datatype *)0 ? 1 : 0) << '\n';
  emitCodeType("w2.d_named",cfOne);

  // d_dedup: identical second decode returns the same container object.
  Datatype *cfOneAgain = decodeTypeStr(factory,&architecture,
    "<type metatype=\"code\" name=\"cf_one\" size=\"1\"/>", err);
  std::cout << "w2.d_dedup.identity=" << (cfOne == cfOneAgain ? 1 : 0) << '\n';
  emitCodeType("w2.d_dedup",cfOneAgain);

  // d_redefine: same name+id, different size; the completed container type
  // fails compareDependency and the previous definition survives.
  Datatype *cfRedefined = decodeTypeStr(factory,&architecture,
    "<type metatype=\"code\" name=\"cf_one\" size=\"2\"/>", err);
  emitThrown("w2.d_redefine", cfRedefined == (Datatype *)0, err);
  emitCodeType("w2.d_redefine_survivor", factory->findByName("cf_one"));

  // d_clash: a named non-code occupant of the name triggers the
  // findByIdLocal metatype check; the int definition survives.
  Datatype *clashInt = decodeTypeStr(factory,&architecture,
    "<type metatype=\"int\" name=\"clash_t\" size=\"4\"/>", err);
  std::cout << "w2.d_clash_int.thrown=" << (clashInt == (Datatype *)0 ? 1 : 0) << '\n';
  std::cout << "w2.d_clash_int.meta=" << static_cast<int4>(clashInt->getMetatype()) << '\n';
  Datatype *clashCode = decodeTypeStr(factory,&architecture,
    "<type metatype=\"code\" name=\"clash_t\" size=\"1\"/>", err);
  emitThrown("w2.d_clash", clashCode == (Datatype *)0, err);
  const Datatype *clashSurvivor = factory->findByName("clash_t");
  std::cout << "w2.d_clash_survivor.meta=" << static_cast<int4>(clashSurvivor->getMetatype()) << '\n';
  std::cout << "w2.d_clash_survivor.size=" << clashSurvivor->getSize() << '\n';

  // d_ctorflags: TypeFactory::decodeCode is PRIVATE in the locked oracle
  // (type.hh:791) — its only public callers are decodeTypeNoRef
  // (isConstructor/isDestructor both false) and decodeTypeWithCodeFlags
  // (which throws before reaching it, wave 1). The constructor/destructor
  // flag chain therefore has no reachable public success observation in
  // 12.0.4; the w1 flag-indifference records are its full observable set.

  // d_varlength: a varlength attribute folds the size into the id
  // (decodeBasic hashSize) and survives the completion wrapper's flag OR.
  Datatype *cfV = decodeTypeStr(factory,&architecture,
    "<type metatype=\"code\" name=\"cf_v\" size=\"1\" varlength=\"true\"/>", err);
  std::cout << "w2.d_varlength.thrown=" << (cfV == (Datatype *)0 ? 1 : 0) << '\n';
  emitCodeType("w2.d_varlength",cfV);
  Datatype *cfVAgain = decodeTypeStr(factory,&architecture,
    "<type metatype=\"code\" name=\"cf_v\" size=\"1\" varlength=\"true\"/>", err);
  std::cout << "w2.d_varlength_dedup.identity=" << (cfV == cfVAgain ? 1 : 0) << '\n';

  // d_unnamed: anonymous code stub (id 0) canonicalized structurally;
  // repeat decode dedups to the same object.
  Datatype *anon1 = decodeTypeStr(factory,&architecture,
    "<type metatype=\"code\" size=\"1\"/>", err);
  std::cout << "w2.d_unnamed.thrown=" << (anon1 == (Datatype *)0 ? 1 : 0) << '\n';
  emitCodeType("w2.d_unnamed",anon1);
  Datatype *anon2 = decodeTypeStr(factory,&architecture,
    "<type metatype=\"code\" size=\"1\"/>", err);
  std::cout << "w2.d_unnamed_dedup.identity=" << (anon1 == anon2 ? 1 : 0) << '\n';

  // ---------------- Wave 3: pointer->code chain via decodeType ----------------
  Datatype *chainCode = decodeTypeStr(factory,&architecture,
    "<type metatype=\"code\" name=\"cf_chain\" size=\"1\"/>", err);
  std::cout << "w3.chain.thrown=" << (chainCode == (Datatype *)0 ? 1 : 0) << '\n';

  Datatype *plainPtr = decodeTypeStr(factory,&architecture,
    "<type metatype=\"ptr\" size=\"8\">"
    "<type metatype=\"code\" name=\"cf_chain\" size=\"1\"/>"
    "</type>", err);
  std::cout << "w3.ptr.thrown=" << (plainPtr == (Datatype *)0 ? 1 : 0) << '\n';
  std::cout << "w3.ptr.meta=" << static_cast<int4>(plainPtr->getMetatype()) << '\n';
  std::cout << "w3.ptr.size=" << plainPtr->getSize() << '\n';
  std::cout << "w3.ptr.wordsize=" << dynamic_cast<const TypePointer *>(plainPtr)->getWordSize() << '\n';
  std::cout << "w3.ptr.ptrto_identity="
            << (dynamic_cast<const TypePointer *>(plainPtr)->getPtrTo() == chainCode ? 1 : 0) << '\n';

  Datatype *widePtr = decodeTypeStr(factory,&architecture,
    "<type metatype=\"ptr\" size=\"8\" wordsize=\"4\">"
    "<type metatype=\"code\" name=\"cf_chain\" size=\"1\"/>"
    "</type>", err);
  std::cout << "w3.ptr_ws.thrown=" << (widePtr == (Datatype *)0 ? 1 : 0) << '\n';
  std::cout << "w3.ptr_ws.wordsize=" << dynamic_cast<const TypePointer *>(widePtr)->getWordSize() << '\n';
}

} // namespace

int main(int argc,char **argv)

{
  try {
    if (argc != 3)
      throw std::invalid_argument(
          "usage: typefactory_codeflags_decode_1204 SPEC_DIRECTORY BINARY");
    runFixture(argv[1],argv[2]);
    return 0;
  }
  catch(const ghidra::LowlevelError &err) {
    std::cerr << "LowlevelError: " << err.explain << '\n';
  }
  catch(const std::exception &err) {
    std::cerr << "std::exception: " << err.what() << '\n';
  }
  return 1;
}

File: tests/oracle/typefactory_codeflags_decode_1204.rs
// TYPEFACTORY-CODEFLAGS-DECODE-0001: Rust comparand for the locked Ghidra
// 12.0.4 decodeTypeWithCodeFlags / decodeCode oracle fixture. Record order
// and values mirror the C++ fixture byte-for-byte; Arc pointer identity is
// the Rust observation of the factory-owned canonical Datatype pointer
// identity Ghidra compares with ==.
//
// The factory mirrors the production BfdArchitecture state by decoding the
// same <size_alignment_map> the compiler spec installs (the C++ side
// inherits it from architecture.init).

use std::sync::{Arc, RwLock};

use rugra::marshal::{Decoder, Element, IdRegistry, TreeDecoder};
use rugra::type_system::datatype::{Datatype, TypeMetatype};
use rugra::type_system::typefactory::TypeFactory;

fn metatype_number(metatype: TypeMetatype) -> i32 {
    match metatype {
        // Locked oracle type.hh:79-99 numbering.
        TypeMetatype::Void => 17,
        TypeMetatype::Unknown => 15,
        TypeMetatype::Int => 14,
        TypeMetatype::Uint => 13,
        TypeMetatype::Bool => 12,
        TypeMetatype::Code => 11,
        TypeMetatype::Float => 10,
        TypeMetatype::Pointer => 9,
        TypeMetatype::Array => 7,
        other => panic!("fixture received unsupported metatype: {other:?}"),
    }
}

/// Build an XML element tree mirroring the C++ fixture's decode strings.
fn elem_new(name: &str, attrs: &[(&str, &str)]) -> Arc<RwLock<Element>> {
    let mut el = Element::new();
    el.set_name(name);
    for (k, v) in attrs {
        el.add_attribute(k, v);
    }
    Arc::new(RwLock::new(el))
}

fn add_child(parent: &Arc<RwLock<Element>>, child: Arc<RwLock<Element>>) {
    parent.write().unwrap().add_child(child);
}

fn elem_owned(name: &str, attrs: Vec<(String, String)>) -> Arc<RwLock<Element>> {
    let mut el = Element::new();
    el.set_name(name);
    for (k, v) in attrs {
        el.add_attribute(&k, &v);
    }
    Arc::new(RwLock::new(el))
}

/// Wrap element trees into a <root> document and hand back a positioned
/// decoder plus the opened root id.
fn root_decoder(children: Vec<Arc<RwLock<Element>>>) -> (TreeDecoder, u32) {
    let root = elem_new("root", &[]);
    for child in children {
        add_child(&root, child);
    }
    let mut decoder = TreeDecoder::new(root, Arc::new(RwLock::new(IdRegistry)));
    let root_id = decoder.open_element();
    (decoder, root_id)
}

/// Wave-1 probe: decodeTypeWithCodeFlags + the cursor partial state.
fn run_code_flags_case(
    factory: &mut TypeFactory,
    key: &str,
    inner: Vec<Arc<RwLock<Element>>>,
    is_constructor: bool,
    is_destructor: bool,
) {
    let void_elem = elem_new("void", &[]);
    let (mut decoder, root_id) = root_decoder({
        let mut all = inner;
        all.push(void_elem);
        all
    });
    let outcome = factory.decode_type_with_code_flags(&mut decoder, is_constructor, is_destructor);
    let (thrown, err_text) = match outcome {
        Ok(_) => (false, String::new()),
        Err(text) => (true, text),
    };
    println!("{key}.thrown={}", thrown as i32);
    println!("{key}.text={err_text}");
    println!("{key}.peek={}", (decoder.peek_element() != 0) as i32);
    decoder.close_element_skipping(0);
    // Open-matching by element name: the C++ side uses the throwing
    // openElement(ELEM_VOID); the Rust decoder registry and the type
    // system's ElementId table assign ids independently, so the element
    // name is the shared observable.
    let sub_id = decoder.open_element();
    let resume = sub_id != 0
        && decoder
            .element_name(sub_id)
            .is_some_and(|name| name == "void");
    if resume {
        decoder.close_element(sub_id);
    }
    println!("{key}.resume_void={}", resume as i32);
    decoder.close_element(root_id);
}

/// Wave-2/3 probe: decode one inner <type> through decodeType.
fn decode_type_str(
    factory: &mut TypeFactory,
    inner: Arc<RwLock<Element>>,
) -> Result<Arc<Datatype>, String> {
    let (mut decoder, root_id) = root_decoder(vec![inner]);
    let outcome = factory.decode_type(&mut decoder);
    decoder.close_element(root_id);
    outcome
}

fn emit_thrown(key: &str, outcome: &Result<Arc<Datatype>, String>) {
    println!("{key}.thrown={}", outcome.is_err() as i32);
    if let Err(text) = outcome {
        println!("{key}.text={text}");
    }
}

fn emit_code_type(key: &str, dt: &Arc<Datatype>) {
    println!("{key}.name={}", dt.get_name());
    println!("{key}.id={}", dt.get_id());
    println!("{key}.size={}", dt.get_size());
    println!("{key}.meta={}", metatype_number(dt.get_metatype()));
    println!(
        "{key}.incomplete={}",
        (dt.get_flags() & rugra::type_system::datatype::type_flags::TYPE_INCOMPLETE != 0) as i32
    );
    println!(
        "{key}.varlength={}",
        (dt.get_flags() & rugra::type_system::datatype::type_flags::VARLENGTH != 0) as i32
    );
    let proto_present = match dt.as_ref() {
        Datatype::Code(c) => c.proto.is_some(),
        _ => false,
    };
    println!("{key}.proto={}", proto_present as i32);
}

/// Inner XML trees for the wave-1 cases, mirroring the C++ strings 1:1.
fn code_type(attrs: &[(&str, &str)]) -> Arc<RwLock<Element>> {
    elem_new("type", attrs)
}

fn code_type_with_proto(
    attrs: &[(&str, &str)],
    proto_attrs: &[(&str, &str)],
    with_returnsym: bool,
) -> Arc<RwLock<Element>> {
    let ty = code_type(attrs);
    let proto = elem_new("prototype", proto_attrs);
    if with_returnsym {
        add_child(&proto, elem_new("returnsym", &[]));
    }
    add_child(&ty, proto);
    ty
}

fn ptr_type(attrs: &[(&str, &str)], child: Arc<RwLock<Element>>) -> Arc<RwLock<Element>> {
    let ty = elem_new("type", attrs);
    add_child(&ty, child);
    ty
}

fn main() {
    let mut factory = TypeFactory::new(8);
    // Mirror BfdArchitecture's decodeDataOrganization: the production
    // factory carries the compiler spec's size alignment map (type.cc:3251
    // does not wipe alignMap).
    {
        let map = elem_new("size_alignment_map", &[]);
        for (size, alignment) in [(1, 1), (2, 2), (4, 4), (8, 8), (16, 16)] {
            add_child(
                &map,
                elem_owned(
                    "entry",
                    vec![
                        ("size".to_string(), size.to_string()),
                        ("alignment".to_string(), alignment.to_string()),
                    ],
                ),
            );
        }
        let root = elem_new("data_organization", &[]);
        add_child(&root, map);
        let mut decoder = TreeDecoder::new(root, Arc::new(RwLock::new(IdRegistry)));
        let _ = factory.decode_data_organization(&mut decoder);
    }

    // ---------------- Wave 1: decodeTypeWithCodeFlags ----------------
    run_code_flags_case(
        &mut factory,
        "w1.a_plain",
        vec![ptr_type(
            &[("metatype", "ptr"), ("size", "8")],
            code_type(&[("metatype", "code"), ("size", "1")]),
        )],
        true,
        false,
    );
    run_code_flags_case(
        &mut factory,
        "w1.a_varargs",
        vec![ptr_type(
            &[("metatype", "ptr"), ("size", "8")],
            code_type_with_proto(
                &[("metatype", "code"), ("size", "1")],
                &[("model", "__stdcall"), ("dotdotdot", "true")],
                true,
            ),
        )],
        false,
        true,
    );
    run_code_flags_case(
        &mut factory,
        "w1.a_model",
        vec![ptr_type(
            &[("metatype", "ptr"), ("size", "8")],
            code_type_with_proto(
                &[("metatype", "code"), ("size", "1")],
                &[
                    ("model", "unknown_cc"),
                    ("constructor", "true"),
                    ("destructor", "true"),
                ],
                false,
            ),
        )],
        true,
        true,
    );
    run_code_flags_case(
        &mut factory,
        "w1.a_thiscall",
        vec![ptr_type(
            &[("metatype", "ptr"), ("size", "8"), ("wordsize", "1")],
            code_type_with_proto(
                &[("metatype", "code"), ("size", "1")],
                &[("model", "__thiscall")],
                true,
            ),
        )],
        false,
        false,
    );
    run_code_flags_case(
        &mut factory,
        "w1.b_code",
        vec![code_type(&[("metatype", "code"), ("size", "1")])],
        true,
        false,
    );
    run_code_flags_case(
        &mut factory,
        "w1.b_unspec",
        vec![code_type(&[("size", "4")])],
        true,
        true,
    );
    run_code_flags_case(
        &mut factory,
        "w1.c_nosize_named",
        vec![code_type(&[("metatype", "ptr"), ("name", "vp")])],
        true,
        false,
    );
    run_code_flags_case(
        &mut factory,
        "w1.c_nosize_anon",
        vec![code_type(&[("metatype", "ptr")])],
        false,
        true,
    );

    // ---------------- Wave 2: decodeCode via decodeType ----------------
    let cf_one = decode_type_str(
        &mut factory,
        code_type(&[("metatype", "code"), ("name", "cf_one"), ("size", "1")]),
    )
    .expect("cf_one decode");
    println!("w2.d_named.thrown=0");
    emit_code_type("w2.d_named", &cf_one);

    let cf_one_again = decode_type_str(
        &mut factory,
        code_type(&[("metatype", "code"), ("name", "cf_one"), ("size", "1")]),
    )
    .expect("cf_one re-decode");
    println!(
        "w2.d_dedup.identity={}",
        Arc::ptr_eq(&cf_one, &cf_one_again) as i32
    );
    emit_code_type("w2.d_dedup", &cf_one_again);

    let redefine = decode_type_str(
        &mut factory,
        code_type(&[("metatype", "code"), ("name", "cf_one"), ("size", "2")]),
    );
    emit_thrown("w2.d_redefine", &redefine);
    let survivor = factory
        .find_by_name("cf_one")
        .expect("cf_one survivor after redefinition");
    emit_code_type("w2.d_redefine_survivor", &survivor);

    let clash_int = decode_type_str(
        &mut factory,
        code_type(&[("metatype", "int"), ("name", "clash_t"), ("size", "4")]),
    )
    .expect("clash_t int decode");
    println!("w2.d_clash_int.thrown=0");
    println!("w2.d_clash_int.meta={}", metatype_number(clash_int.get_metatype()));
    let clash_code = decode_type_str(
        &mut factory,
        code_type(&[("metatype", "code"), ("name", "clash_t"), ("size", "1")]),
    );
    emit_thrown("w2.d_clash", &clash_code);
    let clash_survivor = factory
        .find_by_name("clash_t")
        .expect("clash_t survivor after redefine attempt");
    println!(
        "w2.d_clash_survivor.meta={}",
        metatype_number(clash_survivor.get_metatype())
    );
    println!("w2.d_clash_survivor.size={}", clash_survivor.get_size());

    // d_ctorflags: TypeFactory::decodeCode is private in the oracle
    // (type.hh:791); the constructor/destructor chain has no reachable
    // public success observation in 12.0.4 (see w1 flag-indifference
    // records). Nothing is emitted here, mirroring the C++ fixture.

    let cf_v = decode_type_str(
        &mut factory,
        code_type(&[
            ("metatype", "code"),
            ("name", "cf_v"),
            ("size", "1"),
            ("varlength", "true"),
        ]),
    )
    .expect("cf_v decode");
    println!("w2.d_varlength.thrown=0");
    emit_code_type("w2.d_varlength", &cf_v);
    let cf_v_again = decode_type_str(
        &mut factory,
        code_type(&[
            ("metatype", "code"),
            ("name", "cf_v"),
            ("size", "1"),
            ("varlength", "true"),
        ]),
    )
    .expect("cf_v re-decode");
    println!(
        "w2.d_varlength_dedup.identity={}",
        Arc::ptr_eq(&cf_v, &cf_v_again) as i32
    );

    let anon1 = decode_type_str(&mut factory, code_type(&[("metatype", "code"), ("size", "1")]))
        .expect("anonymous code decode");
    println!("w2.d_unnamed.thrown=0");
    emit_code_type("w2.d_unnamed", &anon1);
    let anon2 = decode_type_str(&mut factory, code_type(&[("metatype", "code"), ("size", "1")]))
        .expect("anonymous code re-decode");
    println!(
        "w2.d_unnamed_dedup.identity={}",
        Arc::ptr_eq(&anon1, &anon2) as i32
    );

    // ---------------- Wave 3: pointer->code chain via decodeType ----------------
    let chain_code = decode_type_str(
        &mut factory,
        code_type(&[("metatype", "code"), ("name", "cf_chain"), ("size", "1")]),
    )
    .expect("cf_chain decode");
    println!("w3.chain.thrown=0");

    let plain_ptr = decode_type_str(
        &mut factory,
        ptr_type(
            &[("metatype", "ptr"), ("size", "8")],
            code_type(&[("metatype", "code"), ("name", "cf_chain"), ("size", "1")]),
        ),
    )
    .expect("pointer->code decode");
    println!("w3.ptr.thrown=0");
    println!("w3.ptr.meta={}", metatype_number(plain_ptr.get_metatype()));
    println!("w3.ptr.size={}", plain_ptr.get_size());
    let (wordsize, ptrto, ptr_meta) = match plain_ptr.as_ref() {
        Datatype::Pointer(p) => (p.wordsize, p.ptr_to.clone(), plain_ptr.get_metatype()),
        _ => panic!("w3.ptr is not a pointer"),
    };
    let _ = ptr_meta;
    println!("w3.ptr.wordsize={wordsize}");
    println!(
        "w3.ptr.ptrto_identity={}",
        Arc::ptr_eq(&ptrto, &chain_code) as i32
    );

    let wide_ptr = decode_type_str(
        &mut factory,
        ptr_type(
            &[("metatype", "ptr"), ("size", "8"), ("wordsize", "4")],
            code_type(&[("metatype", "code"), ("name", "cf_chain"), ("size", "1")]),
        ),
    )
    .expect("wide pointer->code decode");
    println!("w3.ptr_ws.thrown=0");
    let ws = match wide_ptr.as_ref() {
        Datatype::Pointer(p) => p.wordsize,
        _ => panic!("w3.ptr_ws is not a pointer"),
    };
    println!("w3.ptr_ws.wordsize={ws}");
}
